docs/current_docs/features/visualization.mdx

```diff
@@ -0,0 +1,75 @@
+---
+slug: /features/visualization
+---
+
+# Visualization
+
+Dagger works by building up a DAG of operations and evaluating them, often in parallel. By nature, this is a difficult thing to display. It's easy to show the output of one operation at a time, but as soon as you have parallelism, you'll get jumbled output that's hard to understand.
+
+This quickly creates the need to visualize the pipeline, to understand what's going on. Dagger includes two features for this purpose.
+
+## Terminal UI (TUI)
+
+The Dagger CLI includes a real-time visualization feature called the terminal UI (TUI). It shows a full-blown DAG in a style similar to `git log --graph`, with the current state of your DAG's evaluation and a full snapshot of all output at the end.
+
+The TUI is designed for rapid iteration, in the same way regular scripting works: you run a command, it shows live progress, and then leaves all the output in your terminal scrollback. Here's an example of it in action:
+
+![Dagger TUI](/img/current_docs/features/tui.gif)
+
+The TUI is driven by [OpenTelemetry](https://opentelemetry.io/) and is essentially a live-streaming OpenTelemetry trace visualizer. It represents Dagger API calls as OpenTelemetry spans with special metadata. If user code integrates with OpenTelemetry, related spans will appear in the TUI as first-class citizens.
+
+The TUI renders a tree of Dagger API calls, represented as GraphQL queries. A parent-child relationship means that the child call is being made by the parent call.
+- A red X or green check indicates whether the call succeeded or failed.
+- A call's duration is rendered after its name.
+
+The CLI has tiered verbosity: `-v`, `-vv`, `-vvv` (just like `curl`):
+- `-v` keeps spans visible after they complete, rather than disappearing after a brief pause;
+- `-vv` reveals internal and encapsulated spans;
+- `-vvv` reveals sub-100ms spans.
+
+For additional debugging information, add the `--debug` flag to the `dagger call` command.
+
+## Traces
+
+Dagger Cloud provides Traces, a browser-based interface focused on tracing and debugging Dagger pipeline runs. A Trace represents one invocation of a Dagger Function, run locally or in CI. It contains detailed information about the operations performed by the Dagger Function.
+
+Traces provides a UI to debug what went wrong when something fails, focusing on tracing what happened and when it happened. Traces also helps you optimize your Dagger Functions as you author them, by giving you the detailed step-by-step execution of your pipeline and the associated telemetry. This is equally useful for debugging pipelines running on a developer's local environment and for pipelines in CI environments.
```

Comment:
nit (only opinion): Wonder if this could be a bullet point with two items instead of a paragraph.

Reply:
Done